controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -45,9 +47,28 @@
  */
 @Slf4j
 class ZKStreamMetadataStore extends AbstractStreamMetadataStore {
-    private final ZKStoreHelper storeHelper;
+    @VisibleForTesting
+    /**
+     * This constant defines the size of the block of counter values that will be used by this controller instance.
+     * The controller will try to get current counter value from zookeeper. It then tries to update the value in store
+     * by incrementing it by COUNTER_RANGE. If it is able to update the new value successfully, then this controller
+     * can safely use the block `previous-value-in-store + 1` to `previous-value-in-store + COUNTER_RANGE` No other controller
+     * will use this range for transaction id generation as it will be unique assigned to current controller.
+     * If controller crashes, all unused values go to waste. In worst case we may lose COUNTER_RANGE worth of values everytime
+     * a controller crashes.
+     * Since we use a 96 bit number for our counter, so
+     */
+    static final int COUNTER_RANGE = 10000;
+    private ZKStoreHelper storeHelper;
     private final ConcurrentMap<Integer, PathChildrenCache> bucketCacheMap;
     private final AtomicReference<PathChildrenCache> bucketOwnershipCacheRef;
+    private final Object lock;
+    @GuardedBy("lock")
```

Comment:
If it's Atomic, why also guard it with a lock? You are essentially double-locking every time you access this object, which is wasteful and can later lead to deadlocks.

Reply:
Counter and limit are protected by lock because we want to reset both these fields together atomically. 

The logic here is, each controller takes a unique range of bigLong numbers, say 1000-2000. It performs local increments to this and issues unique ids.
Once it has exhausted its local range, it will do a fetch on a zookeeper node to get a unique "next" range and can get 5000-6000 this time. 

So there are two levels of locking because locks in AtomicBigLong is used to protect against concurrent increments to the local counter.
And the lock in this class is used for resetting the "counter and limit" atomically whenever we exhaust local range.

Reply:
So if you only access these guys under a lock, why bother with `AtomicBigLong`? Just mark `BigLong` as `@NotThreadSafe` and add the methods you need on it.

Reply:
AtomicBigLong is used in `InMemoryStream.java` as with inmemory scheme we dont need to distribute the range with limits. So there is only a counter. Which is AtomicBigLong